### DISCUSSION TITLE
TaskOperator for subscribing to asynchronous messages (pub/sub)

### DISCUSSION THREAD

**Discussion by adithya-ramesh** [apache/airflow#16381] [Q&A]
Hi

I have the following pub-sub use case to solve with Airflow DAGs :
- **Task A** triggers a service via HTTP using the **SimpleHttpOperator**, upon successful invocation, the service returns a MQ topic which needs to be subscribed to. 

- **Task B** needs to **subscribe** to this topic (via **MQ**). The service(it is a **long running process**) **asynchronously** publishes on this topic which the DAG Task waits for and then perform subsequent tasks. 

What are the best ways to do Task B in Airflow?
- Is Sensor the best approach?
- Are there any Operators available to do such asynchronous operations? 
- Or should this be managed pythonically, like using asyncio event loops?

**Comment by adithya-ramesh** [comment]
any support/suggestions on this? @potiuk

**Reply by potiuk** [reply]
I think you need to look it up if your MQ is supported. There is for example GCP PubSub set of operators and sensors https://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/pubsub.html other MQs as well. And you can easily add your own MQ support in custom operators following the same patterns.

Sensor is a good approach which does not keep worker busy (in "reschedule" mode for sensors) - but they might not act immediately when the message is published. You can also run your Sensor in a "regular" mode - i.e. waiting for new message "actively" but this will keep worker (whole process) busy for the time of waiting, so it is resource consuming.

We also have some ideas around more "true" async based approach in the making that will be both fast (immediate reaction) and efficient (not taking up many threads/processes):  "Defferable Async Operators : https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=177050929. That will come in Airflow 2.2 or 2.3 likely.
